diesel/src/pg/expression/array.rs

```diff
@@ -43,20 +42,63 @@ pub struct ArrayLiteral<T, ST> {
 ///     vec![2, 4],
 /// ];
 /// assert_eq!(expected, ids);
+///
+/// let ids = diesel::select(array(users.select(id)))
+///     .first::<Vec<i32>>(connection)?;
+/// assert_eq!(vec![1, 2], ids);
 /// #     Ok(())
 /// # }
 /// ```
+pub fn array<ST, T>(elements: T) -> dsl::array<ST, T>
+where
+    T: IntoArrayExpression<ST>,
+    ST: SqlType + TypedExpressionType,
+{
+    elements.into_array_expression()
+}
+
+/// Return type of [`array(tuple_or_subselect)`](super::dsl::array)
+#[allow(non_camel_case_types)]
 #[cfg(feature = "postgres_backend")]
-pub fn array<ST, T>(elements: T) -> ArrayLiteral<T::Expression, ST>
+pub type array<ST, T> = <T as IntoArrayExpression<ST>>::ArrayExpression;
+
+/// Trait for types which can be converted into an expression of type `Array`
+///
+/// This includes tuples of expressions with the same SQL type, and subselects with a single column.
+#[diagnostic::on_unimplemented(
+    message = "Cannot convert `{Self}` into an expression of type `Array<{ST}>`",
+    note = "`the trait bound `{Self}: IntoArrayExpression<{ST}>` is not satisfied. \
+        (`AsExpressionList` is a deprecated trait alias for `IntoArrayExpression`)"
+)]
+pub trait IntoArrayExpression<ST: SqlType + TypedExpressionType> {
+    /// Type of the expression returned by [AsArrayExpression::as_in_expression]
+    type ArrayExpression: Expression<SqlType = sql_types::Array<ST>>;
+
+    /// Construct the diesel query dsl representation of
+    /// the `ARRAY (values)` clause for the given type
+    fn into_array_expression(self) -> Self::ArrayExpression;
+}
+
+/// Implement as a no-op for things that were already arrays (that is, don't wrap in ARRAY[]).
+impl<ST, T> IntoArrayExpression<ST> for T
 where
-    T: AsExpressionList<ST>,
+    T: AsExpression<sql_types::Array<ST>>,
+    ST: SqlType + TypedExpressionType + 'static,
 {
-    ArrayLiteral {
-        elements: elements.as_expression_list(),
-        _marker: PhantomData,
+    type ArrayExpression = <T as AsExpression<sql_types::Array<ST>>>::Expression;
+
+    fn into_array_expression(self) -> Self::ArrayExpression {
+        <T as AsExpression<sql_types::Array<ST>>>::as_expression(self)
     }
 }
```

Comment:
I'm not sure if we should have this impl. Maybe not, if we want to map as close as SQL as possible and forbid writing `array(array_expr)` for something that's actually going to get serialized as just `array_expr`.
It seems easier to add later on than to remove.